---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)
```

# pedprobr <img src="man/figures/logo.png" align="right" height=140 />

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/pedprobr)](https://CRAN.R-project.org/package=pedprobr)
[![](https://cranlogs.r-pkg.org/badges/grand-total/pedprobr?color=yellow)](https://cran.r-project.org/package=pedprobr)
[![](https://cranlogs.r-pkg.org/badges/last-month/pedprobr?color=yellow)](https://cran.r-project.org/package=pedprobr)
<!-- badges: end -->


## Introduction

The main content of pedprobr is an implementation of the Elston-Stewart algorithm for pedigree likelihoods. It is a reboot of the implementation in [paramlink](https://CRAN.R-project.org/package=paramlink) which is no longer actively developed. 

pedprobr is part of the ped suite, a collection of packages for pedigree analysis in R, based on [pedtools](https://github.com/magnusdv/pedtools) for basic handling of pedigrees and marker data. In particular, pedprobr does much of the hard work in the [forrel](https://github.com/magnusdv/forrel) package for relatedness analysis and forensic pedigree analysis. 

The workhorse of the pedprobr package is the `likelihood()` function, which works in a variety of situations:

* complex inbred pedigrees
* pedigrees with inbred founders
* autosomal and X-linked markers
* a single marker or two linked markers
* markers with mutation models (supported by [pedmut](https://github.com/magnusdv/pedmut))

## Installation
To get the current official version of `pedprobr`, install from CRAN as follows:
```{r, eval = FALSE}
install.packages("pedprobr")
```
Alternatively, you can obtain the latest development version from GitHub:
```{r, eval = FALSE}
# install.packages("devtools") # install devtools if needed
devtools::install_github("magnusdv/pedprobr")
```

## Getting started
```{r}
library(pedprobr)
```

To set up a simple example, we first use `pedtools` utilities to create a pedigree and attach to it a marker object. The marker has alleles `1` and `2`, with frequencies 0.2 and 0.8 respectively, and both brothers are heterozygous.
```{r}
x = nuclearPed(nch = 2)
m = marker(x, '3' = 1:2, '4' = 1:2, alleles = 1:2, afreq = c(0.2, 0.8))

x = addMarkers(x, m) # attach the marker
x
```

```{r eval = FALSE}
plot(x, m)
```

```{r pedplot, echo = FALSE, fig.height=3, fig.width=2.5}
plot(x, m, margins = c(0,1,1,1))
```

The pedigree likelihood, i,.e., the probability of observing these genotypes given the pedigree, may now be obtained as follows:
```{r}
likelihood(x, marker1 = 1)
```

## Genotype probability distributions
Besides `likelihood()` the most important functions in pedprobr are:

* `oneMarkerDistribution()` : for a subset of family members, compute their joint genotype distribution at a single marker
* `twoMarkerDistribution()` : for a single family member, compute the joint genotype distribution at two linked markers

In both cases, the distributions are computed conditionally on any known genotypes at the markers in question.

For an illustration of `oneMarkerDistribution()` we may continue our example from above, and answer the following question: _Conditional on the two heterozygous children, what is the joint distribution for the parents?_ 

The answer is easily found as follows:
```{r}
oneMarkerDistribution(x, ids = 1:2, partialmarker = 1, verbose = F)
```


